pkg/auth/oidc.go

```diff
@@ -0,0 +1,206 @@
+package auth
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"sync"
+	"time"
+
+	authenticationv1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+// Response from Doppler's OIDC token exchange
+type OIDCTokenResponse struct {
+	Token     string    `json:"token"`
+	ExpiresAt time.Time `json:"expires_at"`
+}
+
+// Handle OIDC-based authentication
+type OIDCAuthProvider struct {
+	// Kubernetes client for TokenRequest API
+	kubeClient kubernetes.Interface
+
+	// K8s Service Account Name
+	ServiceAccountName string
+	Namespace          string
+	Audiences          []string
+
+	Host string
+	// Doppler Service Account Identity ID
+	IdentityID        string
+	VerifyTLS         bool
+	ExpirationSeconds int64
+
+	// Token management
+	rwm         sync.RWMutex
+	cachedToken string
+	tokenExpiry time.Time
+}
+
+// Creates a new OIDC authentication provider
+func NewOIDCAuthProvider(
+	kubeClient kubernetes.Interface,
+	serviceAccountName string,
+	namespace string,
+	audiences []string,
+	identityID string,
+	expirationSeconds int64,
+	host string,
+	verifyTLS bool,
+) *OIDCAuthProvider {
+	return &OIDCAuthProvider{
+		kubeClient:         kubeClient,
+		ServiceAccountName: serviceAccountName,
+		Namespace:          namespace,
+		Audiences:          audiences,
+		Host:               host,
+		IdentityID:         identityID,
+		VerifyTLS:          verifyTLS,
+		ExpirationSeconds:  expirationSeconds,
+	}
+}
+
+// Returns a Doppler API token, refreshing if necessary
+func (o *OIDCAuthProvider) GetToken(ctx context.Context) (string, error) {
+	o.rwm.RLock()
+	if o.isTokenValid() {
+		token := o.cachedToken
+		o.rwm.RUnlock()
+		return token, nil
+	}
+	o.rwm.RUnlock()
```

Comment:
nit: use defer instead?
```suggestion
	o.rwm.RLock()
	defer o.rwm.RUnlock()
	if o.isTokenValid() {
		token := o.cachedToken
		return token, nil
	}
```

Reply:
Oh, you're absolute right! Sorry about that